s3store/s3store.go

```diff
@@ -259,11 +256,11 @@ func (store S3Store) WriteChunk(id string, offset int64, src io.Reader) (int64,
 			return bytesUploaded, nil
 		}
 
-		if (size - offset) <= store.MinPartSize {
+		if (size - offset) <= effectivePartSize {
```

Comment:
With this new approach, we are calculating an optimal effectivePartSize, which needs to respect the MinPartSize and MaxPartSize limits set in the S3 API specification and take into account absolute max-parts limit of 10.000 parts. However, after the point of calculation, the min and max values are not relevant for our transfer anymore.

Each part we upload will need to align to effectivePartSize except for the last one, which can and will in most cases be lower than effectivePartSize and can also be lower than MinPartSize. Still, whether it is actually lower than MinPartSize does not have any relevance anymore.

However, the parts uploaded prior to the last one, must never be smaller than effectivePartSize, as we either would have to increase the size of another part to compensate for that, or we would need to upload an additional part in order to get everything transferred within the absolute max-parts limit of 10.000 parts. Otherwise we would have to exceed the absolute parts-max limit of 10.000 parts by at least one.

So after the point of calculation of the optimal part size (to suffice all restrictions in S3 API), MinPartSize and MaxPartSize loose all their relevance, as effectivePartSize does respect these limits anyways.